src/configuration_server.erl

```diff
@@ -0,0 +1,155 @@
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_call(config_file, _From, State = #configuration{}) ->
+    {reply, State#configuration.config_file_path, State};
+handle_call(default_directories, _From, State = #configuration{}) ->
+    {reply, {ok, default_directories()}, State};
+handle_call(_Request, _From, State = #configuration{}) ->
+    {reply, ok, State}.
+
+%% @private
+%% @doc Handling cast messages
+-spec handle_cast(Request :: term(), State :: #configuration{}) ->
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_cast(_Request, State = #configuration{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc Handling all non call/cast messages
+-spec handle_info(Info :: timeout() | term(), State :: #configuration{}) ->
+    {noreply, NewState :: #configuration{}} |
+    {noreply, NewState :: #configuration{}, timeout() | hibernate} |
+    {stop, Reason :: term(), NewState :: #configuration{}}.
+handle_info(_Info, State = #configuration{}) ->
+    {noreply, State}.
+
+%% @private
+%% @doc This function is called by a gen_server when it is about to
+%% terminate. It should be the opposite of Module:init/1 and do any
+%% necessary cleaning up. When it returns, the gen_server terminates
+%% with Reason. The return value is ignored.
+-spec terminate(
+    Reason :: (normal | shutdown | {shutdown, term()} | term()),
+    State :: #configuration{}
+) ->
+    term().
+terminate(_Reason, _State = #configuration{}) ->
+    ok.
+
+%% @private
+%% @doc Convert process state when code is changed
+-spec code_change(
+    OldVsn :: term() | {down, term()},
+    State :: #configuration{},
+    Extra :: term()
+) ->
+    {ok, NewState :: #configuration{}} | {error, Reason :: term()}.
+code_change(_OldVsn, State = #configuration{}, _Extra) ->
+    {ok, State}.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+%% @doc This function returns the predefined configuration directories.
+-spec default_directories() ->
+    [file:filename_all()].
+default_directories() ->
+    ?CONFIG_PATHS.
+
+%% @private
+%% @doc This function searches for a configuration file in the default
+%% directories and uses the first found file.
+-spec config_file_name() ->
+    {ok, FileName :: file:filename_all()} | {error, Reason :: term()}.
+config_file_name() ->
+    CandidateNames =
+        [
+            filename:absname_join(Path, File)
+            || Path <- default_directories(), File <- ?FILE_NAMES
+        ],
+    case [File || File <- CandidateNames, filelib:is_regular(File)] of
+        ConfigFiles when length(ConfigFiles) >= 1 -> {ok, lists:nth(1, ConfigFiles)};
+        _ -> {error, {no_file, "No config file found"}}
+    end.
```

Comment:
```erlang
[FirstFile | _] = [
            filename:absname_join(Path, File) ||
            Path <- default_directories(), File <- ?FILE_NAMES, filelib:is_regular(File)
]
```
This would match for a list, with at least one element, but fail with an exception, if the input is an empty list. To counter this, I could use a try-catch-block and return the `{error, ...}` in case of an empty list. But that would be very ugly, in my opinion.